File: server/replay.py
"""
2048 — Sunucu tarafı oyun tekrarı (replay) DOĞRULAMA motoru.

İstemcideki src/app/logic/replay.ts dosyasının BİREBİR Python eşi.
Tohum + hamle dizisinden oyunu yeniden oynatıp skoru KENDİMİZ hesaplarız;
istemcinin iddia ettiği skor kullanılmaz → skor tablosu hile yapılamaz.

Determinizm için mulberry32, JS'in 32-bit tam sayı semantiğiyle
(Math.imul, >>>, |0) birebir taşınmıştır. Parite test/replay_parity ile
istemci transkriptlerine karşı doğrulanır.
"""

CHANCE_OF_FOUR = 0.1
_MASK = 0xFFFFFFFF

CHAR_MOVE = {"U": "up", "D": "down", "L": "left", "R": "right"}


def _imul(a, b):
    """JS Math.imul: 32-bit'e kırpılmış çarpım (bit deseni)."""
    return (a * b) & _MASK


def mulberry32(seed):
    """JS mulberry32 ile birebir aynı sözde-rastgele üretici (0..1)."""
    a = seed & _MASK

    def rng():
        nonlocal a
        a = (a + 0x6D2B79F5) & _MASK
        t = _imul(a ^ (a >> 15), (1 | a) & _MASK)
        t = (((t + _imul(t ^ (t >> 7), (61 | t) & _MASK)) & _MASK) ^ t) & _MASK
        return ((t ^ (t >> 14)) & _MASK) / 4294967296.0

    return rng


def _empty_cells(grid):
    """Boş hücreler — satır öncelikli (istemci ile aynı sıra)."""
    n = len(grid)
    out = []
    for r in range(n):
        for c in range(n):
            if grid[r][c] == 0:
                out.append((r, c))
    return out


def _spawn(grid, rand):
    """Bir taş üretir (iki rand() çekimi). Boş yoksa hiçbir şey yapmaz."""
    import math

    cells = _empty_cells(grid)
    if not cells:
        return
    r, c = cells[int(math.floor(rand() * len(cells)))]
    grid[r][c] = 4 if rand() < CHANCE_OF_FOUR else 2


def move_grid(grid, direction):
    """Bir yönde kaydırıp birleştirir. Dönen: (yeni_grid, kazanç, degisti)."""
    n = len(grid)
    horizontal = direction in ("left", "right")
    toward_start = direction in ("left", "up")
    nxt = [[0] * n for _ in range(n)]
    gained = 0

    for line in range(n):
        vals = []
        for i in range(n):
            idx = i if toward_start else n - 1 - i
            v = grid[line][idx] if horizontal else grid[idx][line]
            if v != 0:
                vals.append(v)

        merged = []
        merged_flag = False
        for v in vals:
            if merged and not merged_flag and merged[-1] == v:
                merged[-1] *= 2
                gained += merged[-1]
                merged_flag = True
            else:
                merged.append(v)
                merged_flag = False

        for i, mv in enumerate(merged):
            idx = i if toward_start else n - 1 - i
            if horizontal:
                nxt[line][idx] = mv
            else:
                nxt[idx][line] = mv

    moved = any(grid[r][c] != nxt[r][c] for r in range(n) for c in range(n))
    return nxt, gained, moved


def replay_game(seed, moves, size):
    """
    Tohum + hamle dizisini yeniden oynatır.
    Dönen: {"score", "maxTile", "moves", "valid"}.
    Geçersiz karakter veya tahtayı değiştirmeyen hamle → valid=False.
    """
    rand = mulberry32(seed & _MASK)
    grid = [[0] * size for _ in range(size)]

    _spawn(grid, rand)
    _spawn(grid, rand)

    score = 0
    count = 0
    for ch in moves:
        direction = CHAR_MOVE.get(ch)
        if direction is None:
            return _invalid(grid, score, count)
        nxt, gained, moved = move_grid(grid, direction)
        if not moved:
            return _invalid(grid, score, count)
        grid = nxt
        score += gained
        count += 1
        _spawn(grid, rand)

    return {"score": score, "maxTile": _max(grid), "moves": count, "valid": True}


def _invalid(grid, score, moves):
    return {"score": score, "maxTile": _max(grid), "moves": moves, "valid": False}


def _max(grid):
    m = 0
    for row in grid:
        for v in row:
            if v > m:
                m = v
    return m


def daily_seed(day):
    """Gün anahtarından (YYYY-MM-DD) tohum — istemci dailySeed() ile birebir.

    FNV-1a 32-bit. Günlük meydan okumada herkes AYNI tahtayı oynar; sunucu
    tohumu kendisi hesaplar, oyuncu kolay bir tohum seçemez.
    """
    h = 2166136261
    for ch in day:
        h ^= ord(ch)
        h = (h * 16777619) & _MASK
    return h or 1

File: server/replay.ts
// ============================================================
//  2048 — Sunucu tarafı oyun tekrarı (replay) DOĞRULAMA motoru (TypeScript).
//
//  NestJS backend'inin kullanacağı, çerçeveden BAĞIMSIZ port. İstemcideki
//  src/app/logic/replay.ts ve Python eşi server/replay.py ile BİREBİR aynı
//  davranır; parite testi (test_replay_parity.ts) ile 150 istemci
//  transkriptine karşı doğrulanır.
//
//  Tohum + hamle dizisinden oyunu yeniden oynatıp skoru KENDİMİZ hesaplarız;
//  istemcinin iddia ettiği skor kullanılmaz → skor tablosu hile yapılamaz.
//
//  Determinizm için mulberry32, JS'in 32-bit tam sayı semantiğiyle
//  (Math.imul, >>>, |0) çalışır. Angular'a bağımlı DEĞİLDİR (kendi kendine
//  yeter): Direction/CHAR_MOVE/mulberry32 burada gömülüdür.
// ============================================================

const CHANCE_OF_FOUR = 0.1;

export type Direction = 'up' | 'down' | 'left' | 'right';

/** Hamle karakteri → yön. İstemci board-logic.CHAR_MOVE ile birebir. */
const CHAR_MOVE: Record<string, Direction> = {
  U: 'up',
  D: 'down',
  L: 'left',
  R: 'right',
};

export interface ReplayResult {
  /** Hesaplanan skor (birleşen karelerin toplamı). */
  score: number;
  /** Ulaşılan en büyük kare. */
  maxTile: number;
  /** İşlenen (geçerli) hamle sayısı. */
  moves: number;
  /** Transkript geçerli mi? (geçersiz hamle/karakter → false) */
  valid: boolean;
}

/**
 * JS mulberry32 ile birebir aynı sözde-rastgele üretici (0..1).
 * İstemci ai.ts.mulberry32 ve Python replay.mulberry32 ile aynı bit deseni.
 */
export function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return function () {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** Boş hücreler — satır öncelikli (istemci emptyCells ile aynı sıra). */
function emptyCells(grid: number[][]): Array<[number, number]> {
  const n = grid.length;
  const out: Array<[number, number]> = [];
  for (let r = 0; r < n; r++) for (let c = 0; c < n; c++) if (grid[r][c] === 0) out.push([r, c]);
  return out;
}

/** Bir taş üretir (iki rand() çekimi). Boş hücre yoksa hiçbir şey yapmaz. */
function spawn(grid: number[][], rand: () => number): void {
  const cells = emptyCells(grid);
  if (cells.length === 0) return;
  const [r, c] = cells[Math.floor(rand() * cells.length)];
  grid[r][c] = rand() < CHANCE_OF_FOUR ? 4 : 2;
}

/**
 * Bir yönde kaydırıp birleştirir (SAF, değer ızgarası üzerinde).
 * Her kare en fazla bir kez birleşir. Dönen: yeni ızgara + kazanç + değişti mi.
 */
export function moveGrid(
  grid: number[][],
  dir: Direction,
): { grid: number[][]; gained: number; moved: boolean } {
  const n = grid.length;
  const horizontal = dir === 'left' || dir === 'right';
  const towardStart = dir === 'left' || dir === 'up';
  const next = Array.from({ length: n }, () => new Array(n).fill(0));
  let gained = 0;

  for (let line = 0; line < n; line++) {
    // İtilen kenardan içeri doğru sıralı değerler (0'lar atılır)
    const vals: number[] = [];
    for (let i = 0; i < n; i++) {
      const idx = towardStart ? i : n - 1 - i;
      const v = horizontal ? grid[line][idx] : grid[idx][line];
      if (v !== 0) vals.push(v);
    }
    // Birleştir (bir kez)
    const merged: number[] = [];
    let mergedFlag = false;
    for (let i = 0; i < vals.length; i++) {
      if (merged.length > 0 && !mergedFlag && merged[merged.length - 1] === vals[i]) {
        merged[merged.length - 1] *= 2;
        gained += merged[merged.length - 1];
        mergedFlag = true;
      } else {
        merged.push(vals[i]);
        mergedFlag = false;
      }
    }
    // Yerleştir
    for (let i = 0; i < merged.length; i++) {
      const idx = towardStart ? i : n - 1 - i;
      if (horizontal) next[line][idx] = merged[i];
      else next[idx][line] = merged[i];
    }
  }

  // Değişti mi?
  let moved = false;
  for (let r = 0; r < n && !moved; r++)
    for (let c = 0; c < n; c++)
      if (grid[r][c] !== next[r][c]) {
        moved = true;
        break;
      }

  return { grid: next, gained, moved };
}

/**
 * Tohum + hamle dizisini yeniden oynatıp skoru hesaplar.
 * `moves`: "U/D/L/R" karakter dizisi. Geçersiz karakter veya tahtayı
 * değiştirmeyen hamle → valid=false (o ana kadarki skor/kare döner).
 */
export function replayGame(seed: number, moves: string, size: number): ReplayResult {
  const rand = mulberry32(seed >>> 0);
  let grid = Array.from({ length: size }, () => new Array(size).fill(0));

  // Oyun başı: iki taş (dört rand() çekimi)
  spawn(grid, rand);
  spawn(grid, rand);

  let score = 0;
  let count = 0;

  for (const ch of moves) {
    const dir = CHAR_MOVE[ch];
    if (dir === undefined) return invalid(grid, score, count);
    const res = moveGrid(grid, dir);
    if (!res.moved) return invalid(grid, score, count); // sahte/bozuk transkript
    grid = res.grid;
    score += res.gained;
    count++;
    spawn(grid, rand);
  }

  return { score, maxTile: maxOf(grid), moves: count, valid: true };
}

function invalid(grid: number[][], score: number, moves: number): ReplayResult {
  return { score, maxTile: maxOf(grid), moves, valid: false };
}

function maxOf(grid: number[][]): number {
  let m = 0;
  for (const row of grid) for (const v of row) if (v > m) m = v;
  return m;
}

/**
 * Gün anahtarından (YYYY-MM-DD) tohum — istemci dailySeed() ile birebir.
 * FNV-1a 32-bit. Günlük meydan okumada herkes AYNI tahtayı oynar; sunucu
 * tohumu kendisi hesaplar, oyuncu kolay bir tohum seçemez.
 */
export function dailySeed(day: string): number {
  let h = 2166136261;
  for (const ch of day) {
    h ^= ch.charCodeAt(0);
    h = Math.imul(h, 16777619);
  }
  return (h >>> 0) || 1;
}
